generated/nidigital/session.py

```diff
@@ -2750,6 +2750,48 @@ def self_test(self):
             raise errors.SelfTestError(code, msg)
         return None
 
+    @ivi_synchronized
+    def write_source_waveform_site_unique(self, waveform_name, waveform_data):
+        '''write_source_waveform_site_unique
+
+        TBD
+
+        Args:
+            waveform_name (str):
+
+            waveform_data ({ site: data, site: data, ... }): Dictionary where each key is the site number and the value is array.array of unsigned int
+
+        '''
+        if len(waveform_data) == 0:
+            # Nothing to do
+            return
+
+        site_list = []
+        # We assume all the entries are the same length (we'll check later) to make the array the correct size
+        # Get an entry from the dictionary from https://stackoverflow.com/questions/30362391/how-do-you-find-the-first-key-in-a-dictionary
+        actual_samples_per_waveform = len(waveform_data[next(iter(waveform_data))])
+        data = array.array('L', [0] * (len(waveform_data) * actual_samples_per_waveform))
+        mv = memoryview(data)
+
+        i = 0
+        for site in waveform_data:
+            if len(waveform_data[site]) != actual_samples_per_waveform:
+                raise ValueError('Mismatched length of waveforms. All must be the same length.')
+            if waveform_data[site].typecode != 'L':
```

Comment:
This goes against ducktyping pretty strongly right?
It forces clients to use array.array even though they may have numpy.array or even a plain old Python list. Seems like something we shouldn't enforce so strictly.

Reply:
We should have system tests for these combinations.

Reply:
Because we wrap the array directly without copying, we would need to explicitly handle the other types. When it comes to ctypes, duck typing doesn't really work.

Reply:
> We should have system tests for these combinations.

I'm working on them. You mean combinations of different values for number of sites and actual_samples_per_waveform?

Reply:
Do we want to explicitly support other types?

Reply:
> When it comes to ctypes, duck typing doesn't really work.

Sure. But that's our problem. We've done work so that this isn't the client's problem before. NI-FGEN is an example of this.

> I'm working on them. You mean combinations of different values for number of sites and actual_samples_per_waveform?

No, I mean client passing different things for the data, such as array.array, list of int, numpy.array.

> Do we want to explicitly support other types?

Is it reasonable to force clients to use array.array? That doesn't seem pythonic to me. So I _think_ we do.


How about this:

* @texasaggie97 Open an issue that these functions should support data as things other than array.array
* We can discuss there and not hold this PR
* @sbethur write the test but have it disabled and point to the issue.

Reply:
#1132 created